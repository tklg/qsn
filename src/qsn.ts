import type { JsonLeafValue, JsonValue } from "./types";

const ESCAPE_CHAR = "!";
const SEPARATOR_CHAR = ",";
const OBJECT_KV_CHAR = ":";
const COLLECTION_START_CHAR = "[";
const COLLECTION_END_CHAR = "]";
const QUOTE_CHAR = "'";
const ESCAPABLE_CHARS = [
  ESCAPE_CHAR,
  SEPARATOR_CHAR,
  OBJECT_KV_CHAR,
  COLLECTION_START_CHAR,
  COLLECTION_END_CHAR,
  QUOTE_CHAR,
];
const REGEX_SPECIAL_CHARS = ["[", "("];

const isLeafValue = (value: unknown): value is JsonLeafValue => {
  return value === null || ["string", "number", "boolean"].includes(typeof value);
};

const escapeString = (str: string): string => {
  if (str === "") {
    return "''";
  }

  const reg = new RegExp(
    `(${ESCAPABLE_CHARS.map((c) => (REGEX_SPECIAL_CHARS.includes(c) ? `\\${c}` : c)).join("|")})`,
    "g",
  );
  return str.replace(reg, "!$1");
};

const unescapeString = (str: string): string => {
  if (str === "''") {
    return "";
  }

  const reg = new RegExp(
    `${ESCAPE_CHAR}(${ESCAPABLE_CHARS.map((c) => (REGEX_SPECIAL_CHARS.includes(c) ? `\\${c}` : c)).join("|")})`,
    "g",
  );
  return str.replace(reg, "$1");
};

const encodeLeafValue = (value: JsonLeafValue): string => {
  if (value === null) {
    return `${ESCAPE_CHAR}n`;
  }

  if (typeof value === "boolean") {
    return `${ESCAPE_CHAR}${value.toString()[0]}`;
  }

  if (typeof value === "number") {
    return `${ESCAPE_CHAR}${value}`;
  }

  return escapeString(value);
};

const decodeLeafValue = (value: string): JsonLeafValue => {
  switch (value) {
    case `${ESCAPE_CHAR}n`:
      return null;
    case `${ESCAPE_CHAR}t`:
      return true;
    case `${ESCAPE_CHAR}f`:
      return false;
    default: {
      if (new RegExp(`^${ESCAPE_CHAR}-?\\d+(\\.\\d+)?(e[+-]\\d+)?$`).test(value)) {
        return Number.parseFloat(value.slice(ESCAPE_CHAR.length));
      }

      return unescapeString(value);
    }
  }
};

const encode = (json: JsonValue): string => {
  if (json === undefined) {
    return "";
  }

  if (isLeafValue(json)) {
    return encodeLeafValue(json);
  }

  if (Array.isArray(json)) {
    return `${COLLECTION_START_CHAR}${json.map(encode).join(SEPARATOR_CHAR)}${COLLECTION_END_CHAR}`;
  }

  if (typeof json === "object") {
    if (Object.keys(json).length === 0) {
      return `${COLLECTION_START_CHAR}${OBJECT_KV_CHAR}${COLLECTION_END_CHAR}`;
    }

    const encodedObject = Object.entries(json)
      .map(([k, v]) => {
        if (typeof k !== "string") {
          throw new Error("Keys must be strings.");
        }

        return `${escapeString(k)}${OBJECT_KV_CHAR}${encode(v)}`;
      })
      .join(SEPARATOR_CHAR);

    return `${COLLECTION_START_CHAR}${encodedObject}${COLLECTION_END_CHAR}`;
  }

  throw new Error(`Unsupported value type: ${typeof json}`);
};

const decode = <T extends JsonValue = JsonValue>(encoded: string): T => {
  if (encoded.startsWith(ESCAPE_CHAR)) {
    return decodeLeafValue(encoded) as T;
  }

  if (encoded.startsWith(COLLECTION_START_CHAR)) {
    // Remove the surrounding brackets
    const inner = encoded.slice(1, -1);

    if (inner === OBJECT_KV_CHAR) {
      return {} as T;
    }

    if (inner === "") {
      return [] as unknown as T; // empty array
    }

    // Helper to check if a character at pos is escaped
    const isEscaped = (str: string, pos: number) => {
      let count = 0;
      for (let i = pos - 1; i >= 0 && str[i] === ESCAPE_CHAR; i--) {
        count++;
      }
      return count % 2 === 1;
    };

    // Find first unescaped separator or colon
    const findUnescaped = (char: string) => {
      for (let i = 0; i < inner.length; i++) {
        if (inner[i] === char && !isEscaped(inner, i)) {
          return i;
        }
      }
      return -1;
    };

    // [[:],[:]]
    // [:],[:]
    const firstUnescapedSep = findUnescaped(SEPARATOR_CHAR);
    const firstUnescapedColon = findUnescaped(OBJECT_KV_CHAR);

    if (
      !inner.startsWith(COLLECTION_START_CHAR) && // If the inner part also starts with "[", it is in an array
      firstUnescapedColon !== -1 &&
      (firstUnescapedSep === -1 || firstUnescapedColon < firstUnescapedSep)
    ) {
      // Object
      const result: Record<string, JsonValue> = {};
      let i = 0;
      while (i < inner.length) {
        // Find key
        let keyEnd = i;
        while (
          keyEnd < inner.length &&
          !(inner[keyEnd] === OBJECT_KV_CHAR && !isEscaped(inner, keyEnd))
        ) {
          keyEnd++;
        }
        let key = inner.slice(i, keyEnd);
        key = unescapeString(key);

        // Find value (may be nested)
        const valueStart = keyEnd + 1;
        let depth = 0;
        let valueEnd = valueStart;
        while (valueEnd < inner.length) {
          const char = inner[valueEnd];
          if (char === COLLECTION_START_CHAR && !isEscaped(inner, valueEnd)) {
            depth++;
          }
          if (char === COLLECTION_END_CHAR && !isEscaped(inner, valueEnd)) {
            depth--;
          }
          if (depth === 0 && inner[valueEnd] === SEPARATOR_CHAR && !isEscaped(inner, valueEnd)) {
            break;
          }
          valueEnd++;
        }
        const value = inner.slice(valueStart, valueEnd);
        result[key] = decode(value);
        // Move to next
        i = valueEnd + 1;
      }
      return result as T;
    }

    // Array
    const items: JsonValue[] = [];
    let i = 0;
    while (i < inner.length) {
      let depth = 0;
      let itemEnd = i;
      while (itemEnd < inner.length) {
        const char = inner[itemEnd];
        if (char === COLLECTION_START_CHAR && !isEscaped(inner, itemEnd)) {
          depth++;
        }
        if (char === COLLECTION_END_CHAR && !isEscaped(inner, itemEnd)) {
          depth--;
        }
        if (depth === 0 && inner[itemEnd] === SEPARATOR_CHAR && !isEscaped(inner, itemEnd)) {
          break;
        }
        itemEnd++;
      }
      const item = inner.slice(i, itemEnd);
      items.push(decode(item));
      i = itemEnd + 1;
    }

    return items as T;
  }

  return decodeLeafValue(encoded) as T;
};

export const QSN = { stringify: encode, parse: decode };
