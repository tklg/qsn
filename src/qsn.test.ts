import { QSN } from "./qsn";

describe("QSN", () => {
  describe("stringify and parse", () => {
    it("should encode and decode null", () => {
      const value: null = null;
      const encoded = QSN.stringify(value);
      expect(encoded).toBe("!n");
      expect(QSN.parse(encoded)).toBe(null);

      expect(encoded).toMatchSnapshot();
    });

    it("should encode and decode booleans", () => {
      expect(QSN.parse(QSN.stringify(true))).toBe(true);
      expect(QSN.parse(QSN.stringify(false))).toBe(false);

      expect(QSN.stringify(true)).toMatchSnapshot();
      expect(QSN.stringify(false)).toMatchSnapshot();
    });

    it("should encode and decode numbers", () => {
      expect(QSN.parse(QSN.stringify(42))).toBe(42);
      expect(QSN.parse(QSN.stringify(-3.14))).toBe(-3.14);

      expect(QSN.stringify(42)).toMatchSnapshot();
      expect(QSN.stringify(-3.14)).toMatchSnapshot();
    });

    it("should encode and decode strings", () => {
      expect(QSN.parse(QSN.stringify("hello"))).toBe("hello");
      expect(QSN.parse(QSN.stringify("!,:["))).toBe("!,:[");

      expect(QSN.stringify("hello")).toMatchSnapshot();
      expect(QSN.stringify("!,:[")).toMatchSnapshot();
    });

    it("should encode and decode arrays", () => {
      const arr = [1, "two", false, null];
      const encoded = QSN.stringify(arr);
      expect(QSN.parse(encoded)).toEqual(arr);
      expect(encoded).toMatchSnapshot();
    });

    it("should encode and decode nested arrays", () => {
      const arr = [1, [2, [3, null]], "test"];
      const encoded = QSN.stringify(arr);
      expect(QSN.parse(encoded)).toEqual(arr);
      expect(encoded).toMatchSnapshot();
    });

    it("should encode and decode objects", () => {
      const obj = { a: 1, b: "two", c: false, d: null };
      const encoded = QSN.stringify(obj);
      expect(QSN.parse(encoded)).toEqual(obj);
      expect(encoded).toMatchSnapshot();
    });

    it("should encode and decode nested objects", () => {
      const obj = { a: { b: { c: [1, 2, 3] } }, d: "test" };
      const encoded = QSN.stringify(obj);
      expect(QSN.parse(encoded)).toEqual(obj);
      expect(encoded).toMatchSnapshot();
    });

    it("should encode and decode complex structures", () => {
      const value = {
        arr: [1, { foo: "bar", baz: [true, false, null] }],
        str: "hello,world!",
        num: 123,
        bool: true,
        nil: null,
      };
      const encoded = QSN.stringify(value);
      expect(QSN.parse(encoded)).toEqual(value);
      expect(encoded).toMatchSnapshot();
    });

    it("should handle empty array and object", () => {
      expect(QSN.parse(QSN.stringify([]))).toEqual([]);
      expect(QSN.parse(QSN.stringify({}))).toEqual({});
      expect(QSN.stringify([])).toMatchSnapshot();
      expect(QSN.stringify({})).toMatchSnapshot();
    });

    it("should escape and unescape special characters in keys and values", () => {
      const obj = { "!key,:[": "!value,:[" };
      const encoded = QSN.stringify(obj);
      expect(QSN.parse(encoded)).toEqual(obj);
      expect(encoded).toMatchSnapshot();
    });
  });
});
