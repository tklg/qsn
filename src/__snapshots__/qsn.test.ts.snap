// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QSN > stringify and parse > should encode and decode arrays -  1`] = `"()"`;

exports[`QSN > stringify and parse > should encode and decode arrays -  2`] = `"(!n)"`;

exports[`QSN > stringify and parse > should encode and decode arrays - 0 1`] = `"((!0))"`;

exports[`QSN > stringify and parse > should encode and decode arrays - 0 2`] = `"(((((!0)))))"`;

exports[`QSN > stringify and parse > should encode and decode arrays - 0,0,0 1`] = `"(((((!0),!0)),!0))"`;

exports[`QSN > stringify and parse > should encode and decode arrays - 1,2,3, 1`] = `"(!1,(!2,(!3,!n)))"`;

exports[`QSN > stringify and parse > should encode and decode arrays - 1,two,false, 1`] = `"(!1,two,!f,!n)"`;

exports[`QSN > stringify and parse > should encode and decode booleans 1`] = `"!t"`;

exports[`QSN > stringify and parse > should encode and decode booleans 2`] = `"!f"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - , 1`] = `"((),())"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - ,[object Object] 1`] = `"((),(:))"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - [object Object] 1`] = `"(([]:{}))"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - [object Object], 1`] = `"((:),())"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - [object Object],[object Object] 1`] = `"((:),(:))"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - { foo: [ 2, { bar: [ 4, { baz: [ 6, { 'deep enough': 7 } ] } ] } ] } 1`] = `"(foo:(!2,(bar:(!4,(baz:(!6,(deep enough:!7)))))))"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - { num: 1, alpha: 'abc', ignore: 'me', change: 'to a function', toUpper: true, obj: { nested_num: 50, alpha: 'abc', nullable: null }, arr: [ 1, 7, 2 ] } 1`] = `"(num:!1,alpha:abc,ignore:me,change:to a function,toUpper:!t,obj:(nested_num:!50,alpha:abc,nullable:!n),arr:(!1,!7,!2))"`;

exports[`QSN > stringify and parse > should encode and decode complex structures - JSON makeTest Pattern pass1,[object Object],[object Object],,-42,true,false,,[object Object],0.5,98.6,99.44,1066,10,1,0.1,1,2,2,beans 1`] = `
"(JSON makeTest Pattern pass1,(object with 1 member:(array with 1 element)),(:),(),!-42,!t,!f,!n,(integer:!1234567890,real:!-9876.54321,e:!1.23456789e-13,E:!1.23456789e+34,'':!2.3456789012e+76,zero:!0,one:!1,space: ,quote:",backslash:\\,controls:

	,slash:/ & /,alpha:abcdefghijklmnopqrstuvwyz,ALPHA:ABCDEFGHIJKLMNOPQRSTUVWYZ,digit:0123456789,0123456789:digit,special:\`1~!!@#$%^&*!(!)_+-={!'!:[!,]}|;.</>?,hex:ģ䕧覫췯ꯍ,true:!t,false:!f,null:!n,array:(),object:(:),address:50 St. James Street,url:http!://www.JSON.org/,comment:// /* <!!-- --,# -- --> */: , s p a c e d :(!1,!2,!3,!4,!5,!6,!7),compact:(!1,!2,!3,!4,!5,!6,!7),jsontext:{"object with 1 member"!:["array with 1 element"]},quotes:&#34; " %22 0x22 034 &#x22;,/\\"쫾몾ꮘﳞ볚

	\`1~!!@#$%^&*!(!)_+-=[]{}|;!:!'!,./<>?:A key can be any string),!0.5,!98.6,!99.44,!1066,!10,!1,!0.1,!1,!2,!2,beans)"
`;

exports[`QSN > stringify and parse > should encode and decode null 1`] = `"!n"`;

exports[`QSN > stringify and parse > should encode and decode numbers - -42 1`] = `"!-42"`;

exports[`QSN > stringify and parse > should encode and decode numbers - -9876.54321 1`] = `"!-9876.54321"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 0 1`] = `"!0"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 0.1 1`] = `"!0.1"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 0.5 1`] = `"!0.5"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 1 1`] = `"!1"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 1 2`] = `"!1"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 1 3`] = `"!1"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 1.23456789e-13 1`] = `"!1.23456789e-13"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 2 1`] = `"!2"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 2 2`] = `"!2"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 2.3456789012e+76 1`] = `"!2.3456789012e+76"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 10 1`] = `"!10"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 98.6 1`] = `"!98.6"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 99.44 1`] = `"!99.44"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 1066 1`] = `"!1066"`;

exports[`QSN > stringify and parse > should encode and decode numbers - 1234567890 1`] = `"!1234567890"`;

exports[`QSN > stringify and parse > should encode and decode objects - { '': '' } 1`] = `"('':'')"`;

exports[`QSN > stringify and parse > should encode and decode objects - { a: 1, b: 'two', c: false, d: null } 1`] = `"(a:!1,b:two,c:!f,d:!n)"`;

exports[`QSN > stringify and parse > should encode and decode objects - {} 1`] = `"(:)"`;

exports[`QSN > stringify and parse > should encode and decode strings - 	 1`] = `"	"`;

exports[`QSN > stringify and parse > should encode and decode strings -   1`] = `" "`;

exports[`QSN > stringify and parse > should encode and decode strings -  1`] = `"''"`;

exports[`QSN > stringify and parse > should encode and decode strings - " 1`] = `"""`;

exports[`QSN > stringify and parse > should encode and decode strings - # -- --> */ 1`] = `"# -- --> */"`;

exports[`QSN > stringify and parse > should encode and decode strings - & 1`] = `"&"`;

exports[`QSN > stringify and parse > should encode and decode strings - &#34; " %22 0x22 034 &#x22; 1`] = `"&#34; " %22 0x22 034 &#x22;"`;

exports[`QSN > stringify and parse > should encode and decode strings - / & / 1`] = `"/ & /"`;

exports[`QSN > stringify and parse > should encode and decode strings - / 1`] = `"/"`;

exports[`QSN > stringify and parse > should encode and decode strings - // /* <!-- -- 1`] = `"// /* <!!-- --"`;

exports[`QSN > stringify and parse > should encode and decode strings - /\\"쫾몾ꮘﳞ볚	\`1~!@#$%^&*()_+-=[]{}|;:',./<>? 1`] = `"/\\"쫾몾ꮘﳞ볚	\`1~!!@#$%^&*!(!)_+-=[]{}|;!:!'!,./<>?"`;

exports[`QSN > stringify and parse > should encode and decode strings - ; 1`] = `";"`;

exports[`QSN > stringify and parse > should encode and decode strings - @ 1`] = `"@"`;

exports[`QSN > stringify and parse > should encode and decode strings - @:0&@:0&@:0&:0 1`] = `"@!:0&@!:0&@!:0&!:0"`;

exports[`QSN > stringify and parse > should encode and decode strings - \\ 1`] = `"\\"`;

exports[`QSN > stringify and parse > should encode and decode strings - \`1~!@#$%^&*()_+-={':[,]}|;.</>? 1`] = `"\`1~!!@#$%^&*!(!)_+-={!'!:[!,]}|;.</>?"`;

exports[`QSN > stringify and parse > should encode and decode strings - {"object with 1 member":["array with 1 element"]} 1`] = `"{"object with 1 member"!:["array with 1 element"]}"`;

exports[`QSN > stringify and parse > should encode and decode strings - | 1`] = `"|"`;

exports[`QSN > stringify and parse > should encode and decode strings - 0123456789 1`] = `"0123456789"`;

exports[`QSN > stringify and parse > should encode and decode strings - ABCDEFGHIJKLMNOPQRSTUVWYZ 1`] = `"ABCDEFGHIJKLMNOPQRSTUVWYZ"`;

exports[`QSN > stringify and parse > should encode and decode strings - abcdefghijklmnopqrstuvwyz 1`] = `"abcdefghijklmnopqrstuvwyz"`;

exports[`QSN > stringify and parse > should encode and decode strings - space character 1`] = `"space character"`;

exports[`QSN > stringify and parse > should encode and decode strings - ģ䕧覫췯ꯍ 1`] = `"ģ䕧覫췯ꯍ"`;

exports[`QSN > stringify and parse > should escape and unescape special characters in keys and values 1`] = `"(!!key!,!:[:!!value!,!:[)"`;
